src/core/server/elasticsearch/client/log_query_and_deprecation.ts

```diff
@@ -0,0 +1,137 @@
+ *
+ * status code
+ * method URL
+ * request body
+ *
+ * so it could be copy-pasted into the Dev console
+ */
+function getResponseMessage(event: DiagnosticResult): string {
+  const errorMeta = getRequestDebugMeta(event);
+  const body = errorMeta.body ? `\n${errorMeta.body}` : '';
+  return `${errorMeta.statusCode}\n${errorMeta.method} ${errorMeta.url}${body}`;
+}
+
+/**
+ * Returns stringified debug information from an Elasticsearch request event
+ * useful for logging in case of an unexpected failure.
+ */
+export function getRequestDebugMeta(event: DiagnosticResult): {
+  url: string;
+  body: string;
+  statusCode: number | null;
+  method: string;
+} {
+  const params = event.meta.request.params;
+  // definition is wrong, `params.querystring` can be either a string or an object
+  const querystring = convertQueryString(params.querystring);
+  return {
+    url: `${params.path}${querystring ? `?${querystring}` : ''}`,
+    body: params.body ? `${ensureString(params.body)}` : '',
+    method: params.method,
+    statusCode: event.statusCode!,
+  };
+}
+
+export const instrumentEsQueryAndDeprecationLogger = ({
+  logger,
+  client,
+  type,
+}: {
+  logger: Logger;
+  client: Client;
+  type: string;
+}) => {
+  const queryLogger = logger.get('query', type);
+  const deprecationLogger = logger.get('deprecation');
+  client.diagnostic.on('response', (error, event) => {
+    if (event) {
+      const opaqueId = event.meta.request.options.opaqueId;
+      const meta = opaqueId
+        ? {
+            http: { request: { id: event.meta.request.options.opaqueId } },
+          }
+        : undefined; // do not clutter logs if opaqueId is not present
+      let queryMsg = '';
+      if (error) {
+        if (error instanceof errors.ResponseError) {
+          queryMsg = `${getResponseMessage(event)} ${getErrorMessage(error)}`;
+        } else {
+          queryMsg = getErrorMessage(error);
+        }
+      } else {
+        queryMsg = getResponseMessage(event);
+      }
+
+      queryLogger.debug(queryMsg, meta);
+
+      if (event.warnings && event.warnings.length > 0) {
+        // Plugins can explicitly mark requests as originating from a user by
+        // removing the `'x-elastic-product-origin': 'kibana'` header that's
+        // added by default. User requests will be shown to users in the
+        // upgrade assistant UI as an action item that has to be addressed
+        // before they upgrade.
+        // Kibana requests will be hidden from the upgrade assistant UI and are
+        // only logged to help developers maintain their plugins
+        const requestOrigin =
+          (event.meta.request.params.headers != null &&
+            (event.meta.request.params.headers[
+              'x-elastic-product-origin'
+            ] as unknown as string)) === 'kibana'
+            ? 'kibana'
+            : 'user';
+
+        // Strip the first 5 stack trace lines as these are irrelavent to finding the call site
```

Comment:
Does it really show a place where a request was initiated? I suspect https://github.com/elastic/kibana/pull/120044/files#diff-b5f83498b6f6c9bb148d4ca6705a2b2a3757d0f8c177ae6a56e8236817eb4702R88 doesn't keep the async stack trace.

Reply:
On main/8.0 yes:

```
Stack trace:
    at Xpack.info (/Users/rudolf/dev/kibana/node_modules/@elastic/elasticsearch/src/api/api/xpack.ts:66:12)
    at LicensingPlugin.fetchLicense (/Users/rudolf/dev/kibana/x-pack/plugins/licensing/server/plugin.ts:180:34)
```

But on 7.16 where it's more critical to trace deprecations it doesn't :(

Reply:
IIRC it does not always work on 8.0 either. But let's keep it for now as is. What are you going to do for `7.16`?

Reply:
I'll remove stack traces for 7.16 since it's never useful, but having the full query is usually enough to track down where the request was made.